### PR TITLE
Make chained calls to flatmap stack safe

### PR DESCRIPTION
**Pull request by derekjw**
Fix for #1700

Uses the same code as andThen, but modified for flatMap

### CODE REVIEW

**Comment by cvogt** [comment]
Looks great :). Thx @derekjw. Let's see if we can get a review from @szeiger

File: slick/src/main/scala/slick/dbio/DBIOAction.scala [hunk 1]
```diff
@@ -554,7 +561,7 @@ object SynchronousDatabaseAction {
     * fused action can fail with a `ClassCastException` during evaluation. */
   private[slick] def fuseUnsafe[R, S <: NoStream, E <: Effect](a: DBIOAction[R, S, E]): DBIOAction[R, S, E] = {
     a match {
-      case FlatMapAction(base: SynchronousDatabaseAction[_, _, _, _], f, ec) if ec eq DBIO.sameThreadExecutionContext =>
+      case FlatMapAction(base: SynchronousDatabaseAction[_, _, _, _], Vector(f), ec) if ec eq DBIO.sameThreadExecutionContext =>
```
Comment (szeiger) [inline slick/src/main/scala/slick/dbio/DBIOAction.scala]: I haven't checked all use cases but this looks suspicious. It won't be able to fuse a `FlatMapAction` that contains multiple mapping functions. Is this good enough? Should it split off the first mapping and only fuse that? Or assume that everything can be fused?

Comment (derekjw) [inline slick/src/main/scala/slick/dbio/DBIOAction.scala]: I'll have a closer look at this. I tried to be conservative here due to the casting, but I think I can improve it.

Comment (derekjw) [inline slick/src/main/scala/slick/dbio/DBIOAction.scala]: I had a closer look, in general it doesn't seem safe to fuse flatMap as we are not able to be sure if the result is fusable, and there are non-database computations going on in the function as well. The [documentation](http://slick.lightbend.com/doc/3.1.0/dbio.html#transactions-and-pinned-sessions) seems to support this as well.

It looks like the only time fuseUnsafe is used internally with transactions? In the cases where this code is triggered during tests it's always with a single chained action.

**Comment by derekjw** [comment]
I found a further problem, if alternating between flatMap and andThen actions we still get a stack overflow. I'm working on a more robust fix, and I'll open a new PR for that.

**Comment by derekjw** [comment]
Created #1703 for a more general solution

**Pull request closed without merging**